Document workaround to disable SNI checks (#834)
stackabletech/nifi-operator: docs/modules/nifi/pages/troubleshooting/index.adoc

# ...
----

For additional details, see https://medium.com/@chnzhoujun/how-to-resolve-sni-issue-when-upgrading-to-nifi-2-0-907e07d465c5[this article].

=== When using Load Balancers or other proxies

When Load Balancers, API Gateways, or other proxy tools sit in front of NiFi, they may forward requests with hostnames that don't match NiFi's TLS certificates.
This commonly occurs because:

* The proxy uses a different hostname than what's in NiFi's certificate
* NiFi is unaware of the external hostnames used to reach the proxy
* Certificate management in Kubernetes makes it complex to include all possible hostnames

==== Workaround: Disable SNI validation

Since this scenario is common and difficult to resolve through certificate management, you can disable SNI validation in NiFi:

[source,yaml]
----
nodes:
  configOverrides:
    nifi.properties:

>       nifi.web.https.sni.required: "false"

Reviewer: I *could* google this, but users likely have the same question, so I think it would be best to answer it in this docs:
I thought it's only a host mismatch problem, isn't only `nifi.web.https.sni.host.check` sufficient? Maybe a short explanation what these two properties do :innocent:

Reviewer: Thanks!

# ...
----

For additional details, see https://medium.com/@chnzhoujun/how-to-resolve-sni-issue-when-upgrading-to-nifi-2-0-907e07d465c5[this article{external-link-icon}^].

=== When using Load Balancers or other proxies

> When Load Balancers, API Gateways, or other proxy tools sit in front of NiFi, they may forward requests with hostnames that don't match NiFi's TLS certificates.

Author: I did not know you can do two suggestions. Neat.

This commonly occurs because:

* The proxy uses a different hostname than what's in NiFi's certificate
* NiFi is unaware of the external hostnames used to reach the proxy
* Certificate management in Kubernetes makes it complex to include all possible hostnames

==== Workaround: Disable SNI validation

> Since this scenario is common and difficult to resolve through certificate management, you can disable SNI validation in NiFi:

Reviewer 2: This is through our own fault (well, secret-operator not supporting external names), so I think we need to revise this:

```suggestion
Since this scenario cannot currently be resolved through certificate management, you can disable SNI validation in NiFi:
```

[source,yaml]
----
nodes:
  configOverrides:
    nifi.properties:
      nifi.web.https.sni.required: "false"
      nifi.web.https.sni.host.check: "false"
----

These properties control NiFi's SNI validation behavior:

> * `nifi.web.https.sni.required`: When `true`, NiFi requires that HTTPS requests include an SNI certificate. Setting to `false` allows connections without SNI certificates.

Reviewer 2: ```suggestion
* `nifi.web.https.sni.required`: When `true`, NiFi requires that HTTPS requests include the SNI extension (containing the `server_name`). Setting to `false` allows connections without SNI certificates.
```

> * `nifi.web.https.sni.host.check`: When `true`, NiFi validates that the SNI certificate's hostname matches the incoming request's hostname. Setting to `false` disables this hostname matching check.

Reviewer 2: We might need to check this: Does the _SNI Check_ check that the `server_name` is the same as the HTTP `Host` name? Or just that it is in the certificate, or both?

There is no SNI Certificate, so I'll rewrite that part, but we still need to ensure it is valid for Jetty 10.

```suggestion
* `nifi.web.https.sni.host.check`: When `true`, NiFi validates that the SNI `server_name` is a valid Subject in the Certificate and that it matches the `host` header of the HTTP request. Setting to `false` disables this hostname matching check.
```

If we aren't sure about the host header matching, then:

```suggestion
* `nifi.web.https.sni.host.check`: When `true`, NiFi validates that the SNI `server_name` is a valid Subject in the Certificate. Setting to `false` disables this hostname matching check.
```

I'm also trying to work out how we can mention Proxy Protocol v2 (PP2) in there... because load balancers can take the original `server_name`, and write it into the Authority header of the PP2. Jetty 10 supports PP2, so in theory it would check both PP2 Authority header (prepended to the TCP stream) or fall back to SNI `server_name` (in TLS Client Hello).

It's hard to validate that until we can have secret-operator add external Subject names. So maybe we can leave that for later.

Reviewer 2: I added a new comment here: https://github.com/stackabletech/nifi-operator/pull/834#discussion_r2378021956

These properties control NiFi's SNI validation behavior:

> * `nifi.web.https.sni.required`: When `true`, NiFi requires that HTTPS requests include the SNI extension (containing the `server_name`). Setting to `false` allows connections without SNI certificates.

Reviewer 2: We might want to mention that when set to true, hostnames must be used, and connections via IP will fail since the SNI extension isn't used with IP.

> * `nifi.web.https.sni.host.check`: When `true`, NiFi validates that the SNI `server_name` is a valid Subject in the Certificate. Setting to `false` disables this hostname matching check.

Reviewer 2: Phew, I'm really glad I wasn't wrong because otherwise I would have been wrong in many recent conversations and would have to rebuild my whole mental model of this.

And yes, the SNI bit is handled at two levels, and I assume Jetty ignore the lower level one by default so that a meaningful response can be sent back to the user in HTTP. The lower layer one is normally for selecting the right certificate when there can be multiple certs (what SNI was designed for).
